vm.c

```diff
@@ -2017,32 +2001,36 @@ hook_before_rewind(rb_execution_context_t *ec, const rb_control_frame_t *cfp,
             break;
           case VM_FRAME_MAGIC_BLOCK:
             if (VM_FRAME_BMETHOD_P(ec->cfp)) {
-                EXEC_EVENT_HOOK(ec, RUBY_EVENT_B_RETURN, ec->cfp->self, 0, 0, 0, frame_return_value(err));
+                VALUE bmethod_return_value = frame_return_value(err);
+                if (THROW_DATA_P(err) && THROW_DATA_STATE(err) == TAG_RETURN) {
+                    // Non-local return terminating at a BMETHOD control frame.
+                    bmethod_return_value = THROW_DATA_VAL(err);
+                }
```

Comment:
I'm not sure about this part, actually. The idea is to fire the event with return value for the last frame that is popped during the unwind process for break/return i.e. the frame for the method that is returning with the RHS of the keyword. This might be firing it for more than just that frame. I will need to check this. 

The caller has this info and could pass it in.